aspects/aspects.go

```diff
@@ -0,0 +1,239 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/jsonutil"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type ErrNotFound struct {
+	Err string
+}
+
+func (e *ErrNotFound) Error() string {
+	return e.Err
+}
+
+func (e *ErrNotFound) Is(err error) bool {
+	_, ok := err.(*ErrNotFound)
+	return ok
+}
+
+var validAccessValues = []string{"read", "write", "read-write"}
+
+// Schema maps dotted paths to underlying storage mechanism.
+type Schema interface {
+	Get(path string, value interface{}) error
+	Set(path string, value interface{}) error
+	Validate(name string) error
+}
+
+// Directory holds a series of related aspects.
+type Directory struct {
+	Name    string
+	schema  Schema
+	aspects map[string]*Aspect
+}
+
+// NewAspectDirectory returns a new aspect directory for the following aspects
+// and views.
+func NewAspectDirectory(name string, mapping map[string]interface{}, schema Schema) (*Directory, error) {
+	a, ok := mapping["aspects"]
+	if !ok {
+		return nil, errors.New("cannot create aspects directory without aspects map")
+	}
+
+	aspects, ok := a.(map[string]interface{})
+	if !ok {
+		return nil, errors.New(`cannot create aspects directory: "aspects" key should map to an aspects map`)
+	}
+
+	aspectDir := Directory{
+		Name:    name,
+		schema:  schema,
+		aspects: make(map[string]*Aspect, len(aspects)),
+	}
+
+	for name, v := range aspects {
+		aspectViews, ok := v.([]map[string]string)
+		if !ok {
+			return nil, errors.New("cannot create aspect: aspect views should be a list of maps")
+		} else if len(aspectViews) == 0 {
+			return nil, errors.New("cannot create aspect without views")
+		}
+
+		aspect := &Aspect{
+			Name:      name,
+			views:     make([]*view, 0, len(aspectViews)),
+			directory: aspectDir,
+		}
+
+		for _, aspectView := range aspectViews {
+			name, ok := aspectView["name"]
+			if !ok || name == "" {
+				return nil, errors.New(`cannot create aspect view without a "name"`)
+			}
+
+			path, ok := aspectView["path"]
+			if !ok || path == "" {
+				return nil, errors.New(`cannot create aspect view without a "path"`)
+			}
+
+			access := aspectView["access"]
+			if access != "" && strutil.ListContains(validAccessValues, access) {
+				return nil, fmt.Errorf("cannot create aspect view: expected \"access\" to be one of %s instead of %q",
+					strutil.Quoted(validAccessValues), access)
+			}
+
+			aspect.views = append(aspect.views, &view{
+				name:   name,
+				path:   path,
+				access: access,
+			})
+		}
+
+		aspectDir.aspects[name] = aspect
+	}
+
+	return &aspectDir, nil
+}
+
+// Aspect return an aspect from the aspect directory.
+func (d *Directory) Aspect(aspect string) *Aspect {
+	return d.aspects[aspect]
+}
+
+// Aspect is a group of aspect views.
+type Aspect struct {
+	Name      string
+	views     []*view
+	directory Directory
+}
+
+func (a *Aspect) Set(view string, value interface{}) error {
+	if err := a.directory.schema.Validate(view); err != nil {
```

Comment:
why this call? given there is no implementation and schema and databag are mixed is a bit unclear